src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/token/PersistentServiceProviderTokenStore.java

```diff
@@ -0,0 +1,422 @@
+        return session != null && currentTimeMillis() > session.getLastRenewalTime() + session.getTimeToLive();
+    }
+
+    private static boolean isTokenExpired(ServiceProviderToken token) {
+        return currentTimeMillis() > token.getCreationTime() + token.getTimeToLive();
+    }
+
+    @Override
+    public Optional<ServiceProviderToken> get(String token) throws StoreException {
+        load();
+        return ofNullable(tokenMap.get(token));
+    }
+
+    @Override
+    public Iterable<ServiceProviderToken> getAccessTokensForUser(String username) {
+        load();
+        return tokenMap.values().stream().filter(token -> equalsIgnoreCase(username, token.getUser())).collect(toList());
+    }
+
+    @Override
+    public ServiceProviderToken put(ServiceProviderToken token) throws StoreException {
+        requireNonNull(token, "token");
+        load();
+        tokenMap.put(token.getToken(), token);
+        save();
+        return token;
+    }
+
+    @Override
+    public void remove(String token) {
+        load();
+        if (tokenMap.remove(token) != null) {
+            save();
+        }
+    }
+
+    @Override
+    public void removeExpiredTokens() throws StoreException {
+        load();
+        boolean needToSave = false;
+        for (ServiceProviderToken token : tokenMap.values()) {
+            if (isTokenExpired(token) && tokenMap.remove(token.getToken()) != null) {
+                needToSave = true;
+            }
+        }
+        if (needToSave) {
+            save();
+        }
+    }
+
+    @Override
+    public void removeExpiredSessions() throws StoreException {
+        load();
+        boolean needToSave = false;
+        for (ServiceProviderToken token : tokenMap.values()) {
+            if (isSessionExpired(token.getSession()) && tokenMap.remove(token.getToken()) != null) {
+                needToSave = true;
+            }
+        }
+        if (needToSave) {
+            save();
+        }
+    }
+
+    @Override
+    public void removeByConsumer(String consumerKey) {
+        load();
+        boolean needToSave = false;
+        for (ServiceProviderToken token : tokenMap.values()) {
+            Consumer consumer = token.getConsumer();
+            if (consumer != null && Objects.equals(consumerKey, consumer.getKey()) &&
+                    tokenMap.remove(token.getToken()) != null) {
+                needToSave = true;
+            }
+        }
+        if (needToSave) {
+            save();
+        }
+    }
+
+    public synchronized void load() {
+        if (tokenMap != null) {
+            return;
+        }
+
+        XmlFile configFile = getConfigFile();
+        if (configFile.exists()) {
+            try {
+                configFile.unmarshal(this);
+            } catch (IOException e) {
+                log.log(SEVERE, "Failed to load OAuth tokens from disk", e);
+                throw new StoreException("Failed to load OAuth tokens");
```

Comment:
Can you please chain the exception?

Reply:
`StoreException` currently doesn't accept a `cause` exception. I thought that was by design for some reason. :)

Reply:
Slightly related: I just noticed `StoreException` is a `RuntimeException`. Why did we explicitly declared it as part of the method signatures in the `ServiceProviderTokenStore`? It has no effect for runtime exceptions.

Reply:
That's strange. I'd argue `StoreException` should be checked, and it should inherit from `Exception` instead. We should force consumers to handle the possibility a store action will fail.

Reply:
Because checked exceptions are bad. You are forcing clients to handle it when they don't know how they will. This results in most of them rethrowing exception to the layer above. A simple task of refactoring causes lots of changes in unintended layers. Hence run time exceptions are used. They are part of method signatures for documentation.

Reply:
+1 to what Gaurav said.